File: Formularios/Agregar/add_comprobante.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using Centrex.Models;

namespace Centrex
{
    public partial class add_comprobante
    {
   public add_comprobante()
 {
            InitializeComponent();
        }

        private void cmd_ok_Click(object sender, EventArgs e)
        {
    if (string.IsNullOrEmpty(txt_comprobante.Text))
   {
        Interaction.MsgBox("El campo 'Comprobante' es obligatorio y está vacio", (MsgBoxStyle)((int)Constants.vbOKOnly + (int)Constants.vbExclamation), "Centrex");
   return;
     }
     else if (cmb_tipoComprobante.SelectedValue is null)
            {
                Interaction.MsgBox("Debe seleccionar un tipo de comprobante", (MsgBoxStyle)((int)Constants.vbOKOnly + (int)Constants.vbExclamation), "Centrex");
                return;
            }
      else if (chk_esMipyme.Checked && cmb_modoMiPyme.SelectedValue is null)
         {
       Interaction.MsgBox("Debe seleccionar un modo MiPyME", (MsgBoxStyle)((int)Constants.vbOKOnly + (int)Constants.vbExclamation), "Centrex");
        return;
            }
  else if (chk_esMipyme.Checked && cmb_comprobanteRelacionado.Text == "Seleccione un comprobante asociado...")
     {
Interaction.MsgBox("El campo 'Comprobant asociado' es obligatorio al ser un comprobante MiPyME y está vacio", (MsgBoxStyle)((int)Constants.vbOKOnly + (int)Constants.vbExclamation), "Centrex");
       return;
      }

            var c = new ComprobanteEntity();

     c.Comprobante = txt_comprobante.Text;
            c.IdTipoComprobante = Convert.ToInt32(cmb_tipoComprobante.SelectedValue);
            c.NumeroComprobante = (int)Math.Round(nup_numero.Value);
            c.PuntoVenta = (int)Math.Round(nup_puntoVenta.Value);
            c.EsFiscal = rb_fiscal.Checked;
  c.EsElectronica = rb_electronico.Checked;
       c.EsManual = rb_manual.Checked;
            c.EsPresupuesto = rb_presupuesto.Checked;
            c.Activo = chk_activo.Checked;
            c.Testing = chk_testing.Checked;
            c.MaxItems = (int?)Math.Round(nup_maxItems.Value);
            c.Contabilizar = chk_contabilizar.Checked;
            c.MueveStock = chk_mueveStock.Checked;
    // .comprobanteRelacionado = cmb_comprobanteRelacionado.SelectedValue
   c.EsMiPyMe = chk_esMipyme.Checked;
    c.CbuEmisor = txt_CBU_emisor.Text;
   c.AliasCbuEmisor = txt_alias_CBU_emisor.Text;
       c.IdModoMiPyme = Convert.ToInt32(cmb_modoMiPyme.SelectedValue);
   c.AnulaMiPyMe = string.IsNullOrEmpty(txt_comprobanteAnulacion.Text)
  ? null
    : txt_comprobanteAnulacion.Text.Substring(0, 1);

            if (VariablesGlobales.edicion == true)
         {
            c.IdComprobante = VariablesGlobales.edita_comprobante.IdComprobante;
     if (comprobantes.updatecomprobante(c) == false)
    {
       Interaction.MsgBox("Hubo un problema al actualizar el comprobante, consulte con su programdor", Constants.vbExclamation);
           closeandupdate(this);
       }
         }
            else
            {
     comprobantes.addcomprobante(c);
      }

  if (chk_secuencia.Checked == true)
    {
          txt_comprobante.Text = "";
 if (VariablesGlobales.id_tipoComprobante_default > 0)
        cmb_tipoComprobante.SelectedValue = VariablesGlobales.id_tipoComprobante_default;
      else
      cmb_tipoComprobante.SelectedIndex = -1;
            nup_numero.Value = 1m;
    nup_puntoVenta.Value = 1m;
    rb_electronico.Checked = true;
                chk_activo.Checked = true;
    chk_testing.Checked = false;
        nup_maxItems.Value = 20m;
           // chk_esMipyme.Checked = False
                // cmb_modoMiPyme.SelectedValue = 1
    cmb_comprobanteRelacionado.Text = "Seleccione un comprobante asociado...";
            }
            else
       {
  closeandupdate(this);
            }
        }

        private void cmd_exit_Click(object sender, EventArgs e)
    {
            closeandupdate(this);
        }

        private void add_comprobante_FormClosed(object sender, FormClosedEventArgs e)
        {
            closeandupdate(this);
      }

        private void add_comprobante_Load(object sender, EventArgs e)
        {
            // Cargo todos los tipos de documentos
   generales.Cargar_Combo(
     ref cmb_tipoComprobante,
    entidad: "TipoComprobanteEntity",
                displaymember: "ComprobanteAfip",
   valuemember: "IdTipoComprobante",
         predet: -1,
          soloActivos: true,
             filtros: null,
       orden: OrdenAsc("ComprobanteAfip"));
            if (VariablesGlobales.id_tipoComprobante_default > 0)
  {
          cmb_tipoComprobante.SelectedValue = VariablesGlobales.id_tipoComprobante_default;
  }
 else
       {
  cmb_tipoComprobante.SelectedIndex = -1;
     cmb_tipoComprobante.Text = "Seleccione un tipo...";
            }

      // Cargo todos los comprobantes asociados
 // cargar_combo(cmb_comprobanteRelacionado, "SELECT id_comprobante, comprobante FROM comprobantes ORDER BY comprobante ASC", basedb, "comprobante", "id_comprobante")
    // cmb_comprobanteRelacionado.Text = "Seleccione un comprobante asociado..."
            // cmb_comprobanteRelacionado.Enabled = False
            // cmb_comprobanteAsociado.SelectedValue = id_tipoComprobante_default

         // Cargo todos los modos de MiPyme
       generales.Cargar_Combo(
 ref cmb_modoMiPyme,
                entidad: "SysModoMiPymeEntity",
              displaymember: "Modo",
    valuemember: "IdModoMiPyme",
         predet: -1,
    soloActivos: false,
       filtros: null,
         orden: OrdenAsc("Modo"));
    if (cmb_modoMiPyme.Items.Count > 0)
            {
              cmb_modoMiPyme.SelectedValue = 1; // 1=No es MiPyME
         }

       rb_electronico.Checked = true;
 chk_activo.Checked = true;

       if (VariablesGlobales.edicion == true | VariablesGlobales.borrado == true)
    {
   chk_secuencia.Enabled = false;
    {
    var withBlock = VariablesGlobales.edita_comprobante;
              txt_comprobante.Text = withBlock.Comprobante;
        cmb_tipoComprobante.SelectedValue = withBlock.IdTipoComprobante;
           cmb_tipoComprobante.Enabled = false;
         nup_numero.Value = withBlock.NumeroComprobante;
        nup_puntoVenta.Value = withBlock.PuntoVenta;
          rb_fiscal.Checked = withBlock.EsFiscal ?? false;
      rb_electronico.Checked = withBlock.EsElectronica ?? false;
  rb_manual.Checked = withBlock.EsManual ?? false;
     rb_presupuesto.Checked = withBlock.EsPresupuesto ?? false;
           chk_activo.Checked = withBlock.Activo;
    chk_testing.Checked = withBlock.Testing;
        nup_maxItems.Value = withBlock.MaxItems ?? 0;
        chk_contabilizar.Checked = withBlock.Contabilizar;
       chk_mueveStock.Checked = withBlock.MueveStock;
               chk_esMipyme.Checked = withBlock.EsMiPyMe;
   // cmb_comprobanteRelacionado.SelectedValue = .comprobanteRelacionado
               txt_CBU_emisor.Text = withBlock.CbuEmisor;
  txt_alias_CBU_emisor.Text = withBlock.AliasCbuEmisor;
   cmb_modoMiPyme.SelectedValue = withBlock.IdModoMiPyme;
         txt_comprobanteAnulacion.Text = withBlock.AnulaMiPyMe;
             }
  }

         if (VariablesGlobales.borrado == true)
  {
           txt_comprobante.Enabled = false;
   cmb_tipoComprobante.Enabled = false;
         nup_numero.Enabled = false;
    nup_puntoVenta.Enabled = false;
          rb_fiscal.Enabled = false;
      rb_electronico.Enabled = false;
     rb_manual.Enabled = false;
     rb_presupuesto.Enabled = false;
       chk_activo.Enabled = false;
      chk_testing.Enabled = false;
    chk_secuencia.Enabled = false;
       cmd_ok.Visible = false;
     cmd_exit.Visible = false;
        chk_contabilizar.Enabled = false;
                chk_mueveStock.Enabled = false;
           chk_esMipyme.Enabled = false;
       // cmb_comprobanteRelacionado.Enabled = False
         txt_CBU_emisor.Enabled = false;
     txt_alias_CBU_emisor.Enabled = false;
             txt_comprobanteAnulacion.Enabled = false;
         cmb_modoMiPyme.Enabled = false;
   Show();
     if (Interaction.MsgBox("¿Está seguro que desea borrar este comprobante?", (MsgBoxStyle)((int)Constants.vbYesNo + (int)Constants.vbQuestion)) == MsgBoxResult.Yes)
   {
         if (comprobantes.borrarcomprobante(VariablesGlobales.edita_comprobante) == false)
           {
          if (Interaction.MsgBox("Ocurrió un error al realizar el borrado del comprobante, ¿desea intectar desactivarlo para que no aparezca en la búsqueda?", (MsgBoxStyle)((int)MsgBoxStyle.Question + (int)MsgBoxStyle.YesNo)) == Constants.vbYes)
          {
       // Realizo un borrado lógico
         if (comprobantes.updatecomprobante(VariablesGlobales.edita_comprobante, true) == true)
  {
     Interaction.MsgBox("Se ha podido realizar un borrado lógico, pero el comprobante no se borró definitivamente." + "\r" + "Esto posiblemente se deba a que el comprobante, tiene operaciones realizadas y por lo tanto no podrá borrarse", Constants.vbInformation);
     }
       else
      {
          Interaction.MsgBox("No se ha podido borrar el comprobante, consulte con el programador");
              }
  }
}
   }
           closeandupdate(this);
       }
        }

        private void cmb_tipoComprobante_KeyPress(object sender, KeyPressEventArgs e)
        {
      // e.KeyChar = ""
        }

     private void chk_esMipyme_CheckedChanged(object sender, EventArgs e)
        {
        bool chk = chk_esMipyme.Checked;

       // cmb_comprobanteRelacionado.Enabled = chk
      txt_CBU_emisor.Enabled = chk;
            txt_alias_CBU_emisor.Enabled = chk;
            txt_comprobanteAnulacion.Enabled = chk;
            cmb_modoMiPyme.Enabled = chk;
        }

    private static List<Tuple<string, bool>> OrdenAsc(string columna) =>
            new List<Tuple<string, bool>> { Tuple.Create(columna, true) };
    }
}

File: __Viejo/Formularios/Agregar/add_comprobante.vb
Public Class add_comprobante

    Private Sub cmd_ok_Click(sender As Object, e As EventArgs) Handles cmd_ok.Click
        If txt_comprobante.Text = "" Then
            MsgBox("El campo 'Comprobante' es obligatorio y está vacio", vbOKOnly + vbExclamation, "Centrex")
            Exit Sub
        ElseIf chk_esMipyme.Checked And cmb_comprobanteRelacionado.Text = "Seleccione un comprobante asociado..." Then
            MsgBox("El campo 'Comprobant asociado' es obligatorio al ser un comprobante MiPyME y está vacio", vbOKOnly + vbExclamation, "Centrex")
            Exit Sub
        End If

        Dim c As New comprobante

        With c
            .comprobante = txt_comprobante.Text
            .id_tipoComprobante = cmb_tipoComprobante.SelectedValue
            .numeroComprobante = nup_numero.Value
            .puntoVenta = nup_puntoVenta.Value
            .esFiscal = rb_fiscal.Checked
            .esElectronica = rb_electronico.Checked
            .esManual = rb_manual.Checked
            .esPresupuesto = rb_presupuesto.Checked
            .activo = chk_activo.Checked
            .testing = chk_testing.Checked
            .maxItems = nup_maxItems.Value
            .contabilizar = chk_contabilizar.Checked
            .mueveStock = chk_mueveStock.Checked
            '.comprobanteRelacionado = cmb_comprobanteRelacionado.SelectedValue
            .esMiPyME = chk_esMipyme.Checked
            .CBU_emisor = txt_CBU_emisor.Text
            .alias_CBU_emisor = txt_alias_CBU_emisor.Text
            .id_modoMiPyme = cmb_modoMiPyme.SelectedValue
            .anula_MiPyME = txt_comprobanteAnulacion.Text
        End With

        If edicion = True Then
            c.id_comprobante = edita_comprobante.id_comprobante
            If updatecomprobante(c) = False Then
                MsgBox("Hubo un problema al actualizar el comprobante, consulte con su programdor", vbExclamation)
                closeandupdate(Me)
            End If
        Else
            addcomprobante(c)
        End If

        If chk_secuencia.Checked = True Then
            txt_comprobante.Text = ""
            cmb_tipoComprobante.SelectedValue = id_tipoComprobante_default
            nup_numero.Value = 1
            nup_puntoVenta.Value = 1
            rb_electronico.Checked = True
            chk_activo.Checked = True
            chk_testing.Checked = False
            nup_maxItems.Value = 20
            'chk_esMipyme.Checked = False
            'cmb_modoMiPyme.SelectedValue = 1
            cmb_comprobanteRelacionado.Text = "Seleccione un comprobante asociado..."
        Else
            closeandupdate(Me)
        End If
    End Sub

    Private Sub cmd_exit_Click(sender As Object, e As EventArgs) Handles cmd_exit.Click
        closeandupdate(Me)
    End Sub

    Private Sub add_comprobante_FormClosed(sender As Object, e As FormClosedEventArgs) Handles Me.FormClosed
        closeandupdate(Me)
    End Sub

    Private Sub add_comprobante_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        'Cargo todos los tipos de documentos
        cargar_combo(cmb_tipoComprobante, "SELECT id_tipoComprobante, comprobante_AFIP FROM tipos_comprobantes ORDER BY comprobante_AFIP ASC", basedb, "comprobante_AFIP", "id_tipoComprobante")
        cmb_tipoComprobante.SelectedValue = id_tipoComprobante_default

        'Cargo todos los comprobantes asociados
        ' cargar_combo(cmb_comprobanteRelacionado, "SELECT id_comprobante, comprobante FROM comprobantes ORDER BY comprobante ASC", basedb, "comprobante", "id_comprobante")
        'cmb_comprobanteRelacionado.Text = "Seleccione un comprobante asociado..."
        'cmb_comprobanteRelacionado.Enabled = False
        'cmb_comprobanteAsociado.SelectedValue = id_tipoComprobante_default

        'Cargo todos los modos de MiPyme
        cargar_combo(cmb_modoMiPyme, "SELECT id_modoMiPyme, modo FROM sys_modoMiPyme ORDER BY id_modoMiPyme ASC", basedb, "modo", "id_modoMiPyme")
        cmb_modoMiPyme.SelectedValue = 1 '1=No es MiPyme

        rb_electronico.Checked = True
        chk_activo.Checked = True

        If edicion = True Or borrado = True Then
            chk_secuencia.Enabled = False
            With edita_comprobante
                txt_comprobante.Text = .comprobante
                cmb_tipoComprobante.SelectedValue = .id_tipoComprobante
                cmb_tipoComprobante.Enabled = False
                nup_numero.Value = .numeroComprobante
                nup_puntoVenta.Value = .puntoVenta
                rb_fiscal.Checked = .esFiscal
                rb_electronico.Checked = .esElectronica
                rb_manual.Checked = .esManual
                rb_presupuesto.Checked = .esPresupuesto
                chk_activo.Checked = .activo
                chk_testing.Checked = .testing
                nup_maxItems.Value = .maxItems
                chk_contabilizar.Checked = .contabilizar
                chk_mueveStock.Checked = .mueveStock
                chk_esMipyme.Checked = .esMiPyME
                'cmb_comprobanteRelacionado.SelectedValue = .comprobanteRelacionado
                txt_CBU_emisor.Text = .CBU_emisor
                txt_alias_CBU_emisor.Text = .alias_CBU_emisor
                cmb_modoMiPyme.SelectedValue = .id_modoMiPyme
                txt_comprobanteAnulacion.Text = .anula_MiPyME
            End With
        End If

        If borrado = True Then
            txt_comprobante.Enabled = False
            cmb_tipoComprobante.Enabled = False
            nup_numero.Enabled = False
            nup_puntoVenta.Enabled = False
            rb_fiscal.Enabled = False
            rb_electronico.Enabled = False
            rb_manual.Enabled = False
            rb_presupuesto.Enabled = False
            chk_activo.Enabled = False
            chk_testing.Enabled = False
            chk_secuencia.Enabled = False
            cmd_ok.Visible = False
            cmd_exit.Visible = False
            chk_contabilizar.Enabled = False
            chk_mueveStock.Enabled = False
            chk_esMipyme.Enabled = False
            'cmb_comprobanteRelacionado.Enabled = False
            txt_CBU_emisor.Enabled = False
            txt_alias_CBU_emisor.Enabled = False
            txt_comprobanteAnulacion.Enabled = False
            cmb_modoMiPyme.Enabled = False
            Me.Show()
            If MsgBox("¿Está seguro que desea borrar este comprobante?", vbYesNo + vbQuestion) = MsgBoxResult.Yes Then
                If (borrarcomprobante(edita_comprobante)) = False Then
                    If (MsgBox("Ocurrió un error al realizar el borrado del comprobante, ¿desea intectar desactivarlo para que no aparezca en la búsqueda?" _
                     , MsgBoxStyle.Question + MsgBoxStyle.YesNo)) = vbYes Then
                        'Realizo un borrado lógico
                        If updatecomprobante(edita_comprobante, True) = True Then
                            MsgBox("Se ha podido realizar un borrado lógico, pero el comprobante no se borró definitivamente." + Chr(13) +
                                "Esto posiblemente se deba a que el comprobante, tiene operaciones realizadas y por lo tanto no podrá borrarse", vbInformation)
                        Else
                            MsgBox("No se ha podido borrar el comprobante, consulte con el programador")
                        End If
                    End If
                End If
            End If
            closeandupdate(Me)
        End If
    End Sub

    Private Sub cmb_tipoComprobante_KeyPress(sender As Object, e As KeyPressEventArgs) Handles cmb_tipoComprobante.KeyPress
        'e.KeyChar = ""
    End Sub

    Private Sub chk_esMipyme_CheckedChanged(sender As Object, e As EventArgs) Handles chk_esMipyme.CheckedChanged
        Dim chk As Boolean = chk_esMipyme.Checked

        'cmb_comprobanteRelacionado.Enabled = chk
        txt_CBU_emisor.Enabled = chk
        txt_alias_CBU_emisor.Enabled = chk
        txt_comprobanteAnulacion.Enabled = chk
        cmb_modoMiPyme.Enabled = chk
    End Sub
End Class